nautobot/extras/models/customfields.py

```diff
@@ -230,11 +230,14 @@ def clean(self):
             except ValidationError as e:
                 raise ValidationError(f"Invalid value for custom field '{field_name}': {e.message}")
 
-        # Check for missing required values
+        # Check for missing values, erroring on required ones and populating non-required ones automatically
         for cf in custom_fields.values():
             # 2.0 TODO: #824 replace cf.name with cf.slug
-            if cf.required and cf.name not in self._custom_field_data:
-                raise ValidationError(f"Missing required custom field '{cf.name}'.")
+            if cf.name not in self._custom_field_data:
+                if cf.required:
+                    raise ValidationError(f"Missing required custom field '{cf.name}'.")
+                else:
+                    self._custom_field_data[cf.name] = cf.default
```

Comment:
I wonder if this should only be done on create and not on update? On the other hand, explicitly deleting a field from `self._custom_field_data` in an update probably **should** revert it to its default value if applicable. So maybe it's fine as-is. :-)

Reply:
Take this for example:

```
# Assume Site has a custom field
site = Site.objects.create(name="Banana", description: "Banana Land")
site.validated_save()
```

If I understand the proposal correctly, then `site.custom_field_data` would be empty after this, which IMO is not desirable.

Reply:
Agreed. This comment was a half-baked thought and I convinced myself that the way you have it is fine by the time I was done writing it. :-)